### DISCUSSION TITLE
mill-buildr: Splitting large `build.sc` files into an scala project itself.

### DISCUSSION THREAD

**Discussion by vic** [com-lihaoyi/mill#2014] [Show and tell]
At work we've been using mill since about late 2018 and never looked back!. Thanks to all people who have made of `mill` an awesome tool.

Our build.sc which was initially tiny grew a lot of modules over time (all in a mono-repo, eg, models, daos, remote-services, web-server, etc). About two years ago we started splitting our build.sc into smaller modules (other `$file` importable `.sc` files) and mill always worked fine but the "`.sc` editing" experience was not that good even using IntelliJ, they certainly looked like scala files but not-so-experienced developers were missing auto-completion and jump-to-definition features from their IDE. At the time I believe there was no bsp integration and all we had was GenIdea, but IntelliJ was not very good at finding references over lots of `.sc` files.

So, it was really important for me to make code on our "build-definitions" editable by anyone on the team. Have them use their IDE with all features while editing our build project, just like editing any other mill-scala project.

Trying to find a solution for this, I found a *hack* (we have been using this since about ~2020 and finally extracted the code to share with other people). The trick is, define a `buildr` project which has all our modules definitions as traits (a regular mill-scala project) and then load the runClasspath into the mill's interp runtime. After that, all `build.sc` file does is just create top level objects, basically just a list of objects extending their particular module implementations. This way the `build.sc` keeps small and clean and all the meat is in the `buildr` project.

So, this `BuildrModule` can be seen more as a pattern than a library, anyways, just shared a repo if other people having huge mill-projects can benefit from this idea:

https://github.com/vic/mill-buildr

(see the example `build.sc` and example `buildr/src/*.scala` on the repo)

**Comment by vic** [comment]
as a totally unrelated side note: we do have a picture of @lihaoyi at work to bless all our developments. his awesome libraries and book has helped us a lot to get over the difficult times. 

![1710ee0f656c866b.jpg](https://user-images.githubusercontent.com/331/188054807-78a80d65-5fce-4a09-b751-2376ac8a254f.jpg)

**Reply by vic** [reply]
The lengend reads: 
  "This home is lihaoyi-ist, we don't take propaganda from other languages without types."
and 
  "Read my messages in the code, there is the truth - lihaoyi, 2018"

**Comment by lefou** [comment]
I remember that we discussed something like this in the chats. Nice it worked out for you. And thank you for sharing!

In the meantime, I experimented with `using directives` as comments, which are use by Scala CLI project. Turns out, they are very nice to declare plugin dependencies and source-includes. And in that experiment, I was also able to include and compile classic scala files to be used in Mill projects. My time is currently limited to continue on that, but I think that's the direction I most likely want to follow in upcoming Mill milestones.
